src/H5RT.c

```diff
@@ -0,0 +1,905 @@
+
+#include "H5RTpkg.h"
+
+/***********/
+/* Headers */
+/***********/
+#include "H5private.h"   /* Generic functions */
+#include "H5Eprivate.h"  /* Error handling */
+#include "H5FLprivate.h" /* Free lists */
+
+H5FL_DEFINE_STATIC(H5RT_t);
+H5FL_DEFINE_STATIC(H5RT_node_t);
+
+static herr_t H5RT__bulk_load(H5RT_node_t *node, int rank, H5RT_leaf_t *leaves, size_t count,
+                              int prev_sort_dim);
+static herr_t H5RT__search_recurse(H5RT_node_t *node, int rank, hsize_t min[], hsize_t max[],
+                                   H5RT_result_set_t *result_set);
+static herr_t H5RT__node_copy(H5RT_node_t *dest_node, const H5RT_node_t *src_node,
+                              const H5RT_leaf_t *old_leaves_base, H5RT_leaf_t *new_leaves_base);
+static void   H5RT__free_recurse(H5RT_node_t *node);
+
+/* Result buffer helper functions */
+static herr_t H5RT__result_set_init(H5RT_result_set_t *result_set);
+static herr_t H5RT__result_set_add(H5RT_result_set_t *result_set, H5RT_leaf_t *leaf);
+static herr_t H5RT__result_set_grow(H5RT_result_set_t *result_set);
+static void   H5RT__result_set_cleanup(H5RT_result_set_t *result_set);
+
+#if defined(H5_HAVE_DARWIN) || defined(H5_HAVE_WIN32_API)
+static int H5RT__leaf_compare(void *dim, const void *leaf1, const void *leaf2);
+#else
+static int H5RT__leaf_compare(const void *leaf1, const void *leaf2, void *dim);
+#endif
+
+/* Check if two hyper-rectangles specified by (min1, max1) and (min2, max2) intersect */
+/*-------------------------------------------------------------------------
+ * Function:    H5RT_leaf_init
+ *
+ * Purpose:     Initialize an R-tree leaf with dynamic coordinate allocation
+ *              based on the specified rank. The leaf structure itself must
+ *              already be allocated by the caller.
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5RT_leaf_init(H5RT_leaf_t *leaf, int rank, void *record)
+{
+    herr_t ret_value = SUCCEED;
+
+    FUNC_ENTER_NOAPI(FAIL)
+
+    if (!leaf)
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "invalid leaf pointer");
+
+    if (rank < 1 || rank > H5S_MAX_RANK)
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "invalid rank");
+
+    /* Zero out the leaf structure */
+    memset(leaf, 0, sizeof(H5RT_leaf_t));
+
+    /* Allocate coordinate arrays as single block: 3 * rank * sizeof(hsize_t) */
+    if (NULL == (leaf->_coords = (hsize_t *)malloc(3 * (size_t)rank * sizeof(hsize_t))))
+        HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL, "failed to allocate leaf coordinates");
+
+    /* Set up pointers to sections of the coordinate block */
+    leaf->min = leaf->_coords;
+    leaf->max = leaf->_coords + rank;
+    leaf->mid = leaf->_coords + (2 * rank);
+
+    leaf->rank   = rank;
+    leaf->record = record;
+
+done:
+    if (ret_value < 0 && leaf && leaf->_coords) {
```

Comment:
Nitpicking a bit: it seems better to make use of `H5RT_leaf_cleanup()` here to keep the cleanup logic consistent.